### PR TITLE
[Refactor][Perf] Replace _LocalPredictorKVCache with SDPA

### PR DESCRIPTION
**Pull request by gcanlin**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

To solve some problems:
- The code predictor (a small 5-layer sub-model for residual codebook prediction) was borrowing model runner v2's attn_utils (vllm.v1.worker.gpu.attn_utils) to manage its own KV cache, but vllm-omni uses model runner v1. The current implementation is so hacky.
- We have to add the NPU-specifc code branch because v2 backends needed different metadata formats.

The code predictor doesn't need paged attention because paged attention solves three problems it doesn't have:

- Variable-length sequences — Paged attention manages memory efficiently when sequences have unpredictable lengths (hundreds to thousands of tokens). The code predictor always runs exactly num_code_groups tokens (~32), known at init time.
- Memory fragmentation across concurrent sequences — Paged attention avoids fragmentation when the engine juggles thousands of sequences with different lifetimes. The code predictor runs a small batch, allocates at startup, and reuses the same buffer every frame.
- Shared KV cache pool managed by the scheduler — The main LLM's KV cache is allocated/freed by vLLM's scheduler across requests. The code predictor's KV cache is private — it's allocated once, used for ~30 decode steps, then zeroed. No scheduler involvement.

## Test Plan

```
CUDA_VISIBLE_DEVICES=0 python -m vllm_omni.entrypoints.cli.main serve     "Qwen/Qwen3-TTS-12Hz-1.7B-CustomVoice"     --omni --host 127.0.0.1 --port 8000     --stage-configs-path benchmarks/qwen3-tts/vllm_omni/configs/qwen3_tts_bs1.yaml     --trust-remote-code
```

```
python benchmarks/qwen3-tts/vllm_omni/bench_tts_serve.py     --port 8000     --num-prompts 50     --max-concurrency 1 4 10     --save-audio     --result-dir benchmarks/qwen3-tts/results/
```

## Test Result

It looks like RTF values have some problem...

Before:

```
======================================================================
 Summary: async_chunk
======================================================================
  Conc   TTFP(ms)     E2E(ms)      RTF      Audio(s/s)   Req/s
  ------ ------------ ------------ -------- ------------ --------
  1      2502.4       7012.5       2.477    0.40         0.1
  4      23747.9      27808.0      9.832    0.41         0.1
  10     70739.9      75256.3      23.191   0.41         0.1
======================================================================
```

After:

```
======================================================================
 Summary: async_chunk
======================================================================
  Conc   TTFP(ms)     E2E(ms)      RTF      Audio(s/s)   Req/s
  ------ ------------ ------------ -------- ------------ --------
  1      2042.6       5650.0       1.996    0.50         0.2
  4      19093.7      23453.3      8.049    0.51         0.2
  10     51751.7      55537.6      17.790   0.51         0.2
======================================================================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 1]
```diff
@@ -1,198 +1,212 @@
...
-            kv_cache_groups=[
-                KVCacheGroupSpec(layer_names=list(kv_cache_spec_by_layer.keys()), kv_cache_spec=merged_spec)
-            ],
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.total_num_heads = num_heads
+        self.num_heads = num_heads
+        self.total_num_kv_heads = num_kv_heads
+        self.num_kv_heads = num_kv_heads
+        self.head_dim = head_dim or hidden_size // num_heads
```
[the remark this answers is not in the document]
Comment (gcanlin) [reply vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:55]: @chatgpt-codex-connector We don't need TP for Qwen3-TTS. It's 1.7B model.

**Comment by gcanlin** [comment]
cc @Sy0307 @linyueqian

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 2]
```diff
@@ -367,92 +394,77 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
+                self._max_seq_len,
+                self._head_dim,
+                dtype=torch.bfloat16,
+                device=device,
+            )
+            v = torch.zeros(
+                batch_size,
+                self._num_kv_heads,
+                self._max_seq_len,
+                self._head_dim,
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:413]: **Hardcoded dtype**

Consider deriving dtype from config or input instead of hardcoding `bfloat16`:

```python
dtype = inputs_embeds.dtype  # or self.config.torch_dtype
k = torch.zeros(..., dtype=dtype, device=device)
v = torch.zeros(..., dtype=dtype, device=device)
```

This ensures consistency if the model runs in a different precision.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 3]
```diff
@@ -367,92 +394,77 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
-            out = self.model(positions=positions, inputs_embeds=hs)
+        positions = torch.arange(qlen, dtype=torch.long, device=device).unsqueeze(0).expand(bsz, -1)
+
+        out = self.model(positions=positions, inputs_embeds=hs, kv_caches=self._kv_caches, seq_len=qlen)
 
         # Gather last token per request.
-        last_idx = torch.arange(qlen - 1, bsz * qlen, step=qlen, device=out.device, dtype=torch.long)
-        last_h = out.index_select(0, last_idx)
+        last_h = out[:, -1, :]  # [B, hidden]
         logits = self.lm_head[0](last_h)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:447]: **Batch size reallocation**

When `bsz` increases, the old cache is discarded and reallocated. This works because reallocation should only happen during prefill (not mid-generation).

Consider adding an assertion or comment documenting this assumption:

```python
# Reallocation only expected during prefill; cache is reset between generations.
if self._kv_caches is None or self._kv_caches[0][0].shape[0] < bsz:
    self._kv_caches = self._allocate_kv_caches(bsz, device)
```

**Comment by Sy0307** [comment]
LGTM. Good Work! I think it's worthwhile to refactor `build_attn_metadata`/ `set_forward_context` here to eliminate the time overhead. I previously had other ideas to reduce the bubbles here — for example, instead of doing incremental KV-cache decoding at each AR step, treating the "entire prefix generated so far (growing sequence)" as a prefill and running it from scratch each time. 

However, after testing #1591, I found that it essentially eliminates the bubbles and redundant Python calls, achieving approximately 20% performance improvement.

**Comment by Sy0307** [comment]
> for example, instead of doing incremental KV-cache decoding at each AR step, treating the "entire prefix generated so far (growing sequence)" as a prefill and running it from scratch each time.

Though I wonder if this approach might perform better when combined with torch.compile? I'm still not sure, but this PR can be completed and merged first, and afterwards I'll try to write an implementation matching the description above to compare the performance — we can have more discussion at that point.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 4]
```diff
@@ -367,92 +394,77 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
+    def _allocate_kv_caches(self, batch_size: int, device: torch.device) -> list[KVCache]:
+        """Allocate dense KV cache tensors for all layers."""
+        caches: list[KVCache] = []
+        for _ in range(self._num_layers):
+            k = torch.zeros(
+                batch_size,
+                self._num_kv_heads,
+                self._max_seq_len,
+                self._head_dim,
+                dtype=torch.bfloat16,
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:406]: Why hardcode bfloat16 here? Shouldn't this come from the model config?
